Refactor chain monitoring

jkczyz opened the pull request:
`ChainWatchInterface` introduces a dependency between `BlockNotifier`'s listeners, which can lead to complex interactions which aren't easy to understand and can be error prone. ~~This PR refactors `BlockNotifier` to take the entire block and leaves filtering to individual `ChainListener`s where required.~~

~~As a result, the `block_connected` method interfaces are identical between `BlockNotifier` and `ChainListener`. Thus, an adaptor is not needed in #614. Further refactoring around `ChainWatchInterface` and the `reentered` logic is left to a follow-up PR.~~

~~The PR also simplifies `BlockNotifier` tests by removing duplication and unnecessary dependencies.~~

This PR refactors chain monitoring in the following manner:
- Add `chain::Access` and `chain::Filter` traits
- Replace `ManyChannelMonitor` with `chain::Watch` trait, which is functionally the same
- Replace `ChainWatchInterface` in `NetGraphMsgHandler` with an optional `chain::Access`
- Replace `ManyChannelMonitor` in `ChannelManager` with `chain::Watch`
- Replace `ChainWatchInterface` in `SimpleManyChannelMonitor` with an optional `chain::Filter`
- Rename `SimpleManyChannelMonitor` as `ChainMonitor`
- Remove `ChainWatchInterface`, `ChainWatchInterfaceUtil`, `ChainWatchedUtil`, and `ChainError`
- Remove `BlockNotifier` and `ChainListener`

This allows for various setups like using a full node, Electrum, or a BIP157/BIP158 light client for chain source, as wells as local or remote channel monitoring.

TheBlueMatt commented:
Wait, how does one go about implementing using this without downloading the full block?

jkczyz commented:
> Wait, how does one go about implementing using this without downloading the full block?

**TL; DR:** Defining an interface in terms of a `ChainListener` trait is only useful if something is making use of the trait.

Chatted a bit offline. My thinking is blocks are fed through `BlockNotifier` to the `ChainListener`s, so there's no reason for the interface to differ.

@TheBlueMatt noted that for bandwidth reasons users may want to feed partial blocks to a `ChainListener`. They would do so by bypassing `BlockNotifier` entirely.

I'd argue that `ChainListener` is only useful in the context of `BlockNotifier`. If a user must feed a partial block to a listener directly, then there is no need to do so through the `ChainListener` interface -- the user could simply call a non-trait method. Such a method should be added as part of each listener's public interface (which I'm happy to do).

Alternatively, a new abstraction can be added to notify listeners of partial blocks or possibly individual transactions. This could use a separate listener interface (e.g., perhaps `TransactionListener` vs `BlockListener`) or expand upon the current `ChainListener` interface by adding another method. Or possibly using some other variation of these.

TheBlueMatt commented:
I’m certainly happy with either - the original goal was a generic interface across potentially many structs that listen for chain data, but after all was said and done we ended up with only two that care - ChannelMonitors, which may need the rescanning logic and a single ChannelManager which (I believe) does not. Given that, it may not make any sense to have a trait at all and just focus on utilities to do the rescanning for ChannelMonitors.

> On Jun 23, 2020, at 13:12, Jeffrey Czyz <notifications@github.com> wrote:
> 
> ﻿
> Wait, how does one go about implementing using this without downloading the full block?
> 
> TL; DR: Defining an interface in terms of a ChainListener trait is only useful if something is making use of the trait.
> 
> Chatted a bit offline. My thinking is blocks are fed through BlockNotifier to the ChainListeners, so there's no reason for the interface to differ.
> 
> @TheBlueMatt noted that for bandwidth reasons users may want to feed partial blocks to a ChainListener. They would do so by bypassing BlockNotifier entirely.
> 
> I'd argue that ChainListener is only useful in the context of BlockNotifier. If a user must feed a partial block to a listener directly, then there is no need to do so through the ChainListener interface -- the user could simply call a non-trait method. Such a method should be added as part of each listener's public interface (which I'm happy to do).
> 
> Alternatively, a new abstraction can be added to notify listeners of partial blocks or possibly individual transactions. This could use a separate listener interface (e.g., perhaps TransactionListener vs BlockListener) or expand upon the current ChainListener interface by adding another method. Or possibly using some other variation of these.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jkczyz commented:
Rebased and resolved conflicts with #638.

From yesterday's discussions, I'm working on adding an interface to `ChannelManager` for partial blocks. Will ping the PR when ready.

There's also a fuzz test breakage at the following line, which I haven't had much time to dig into: https://github.com/rust-bitcoin/rust-lightning/blob/49b8c58165b0e1a6545021a0c91add92f2ebeadd/fuzz/src/full_stack.rs#L898

TheBlueMatt commented:
Hmm, its possible you broke it, its also possible that check is too aggressive and its checking that we hit that line multiple times due to a rescan?

jkczyz commented:
Updated to support both full and partial (filtered) blocks. Figured the cleanness way to support filtered blocks is to define `ChainListener::block_connected` to handle both cases. Looks a little similar to before, only using a vector of tuples rather than two separate vectors. I kept `BlockNotifier`'s interface in terms of `Block`, but this could be changed later if needed.

> Hmm, its possible you broke it, its also possible that check is too aggressive and its checking that we hit that line multiple times due to a rescan? If you compare the test output prior to the patch and after you should be able to figure out if its correct.

Looks like I broke it while updating `MoneyLossDetector::block_connected`. Fixed now after making the change mentioned above.

TheBlueMatt commented on lightning/src/chain/chaininterface.rs:
```diff
@@ -55,7 +55,7 @@ pub trait ChainWatchInterface: Sync + Send {
 	/// Gets the list of transaction indices within a given block that the ChainWatchInterface is
 	/// watching for.
-	fn filter_block(&self, block: &Block) -> Vec<usize>;
+	fn filter_block(&self, header: &BlockHeader, txdata: &[(usize, &Transaction)]) -> Vec<usize>;
```
Why define this in terms of pre-filtered data? Seems like you could just pass in a &[Transaction] or so and then you don't have to do any crazy iterator-collect allocations to call this with a block?

TheBlueMatt replied:
> I would say it supports either filtered or unfiltered blocks rather being defined in terms of pre-filtered data.

Hmm, arguably, but why support anything other than a full block anyway? I'm not sure what te use-case is for filtering a pre-filtered block at all?

jkczyz replied:
That's a fair point. I'm re-thinking this after some of our offline discussions and will have a revised proposal for these interfaces.

TheBlueMatt commented on lightning/src/chain/chaininterface.rs:
```diff
@@ -71,22 +71,10 @@ pub trait BroadcasterInterface: Sync + Send {
-	/// This also means those counting confirmations using block_connected callbacks should watch
-	/// for duplicate headers and not count them towards confirmations!
-	fn block_connected(&self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[usize]);
+	/// Notifies a listener that a block was connected. Transactions may be filtered and are given
```
This doc could use more details, I think (eg parameter usage, etc).

TheBlueMatt commented on lightning/src/chain/chaininterface.rs:
```diff
@@ -71,22 +71,10 @@ pub trait BroadcasterInterface: Sync + Send {
-	fn block_connected(&self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[usize]);
+	/// Notifies a listener that a block was connected. Transactions may be filtered and are given
+	/// paired with their position within the block.
+	fn block_connected(&self, header: &BlockHeader, txdata: &[(usize, &Transaction)], height: u32);
```
Can you elaborate a bit more on the reasoning for the two-slices -> slice-of-pairs change? I think I agree its more of the "correct" API, but dunno if its worth the change which will probably require some work on the bindings end - did you have a specific reasoning in mind or just think its cleaner?

jkczyz replied:
Partly to make it cleaner and consistent throughout. Also less error prone since listeners don't need to manually match up transactions and indexes. Ideally, the indexes are not needed. Right now it seems specific to `ChannelManager`, but perhaps there are other uses that I haven't thought of.

valentinewallace commented on lightning/src/ln/channelmonitor.rs:
```diff
@@ -183,27 +183,32 @@ impl<Key : Send + cmp::Eq + hash::Hash, ChanSigner: ChannelKeys, T: Deref + Sync
-					for (idx, output) in outputs.iter().enumerate() {
-						self.chain_monitor.install_watch_outpoint((txid.clone(), idx as u32), &output.script_pubkey);
+	fn block_connected(&self, header: &BlockHeader, txdata: &[(usize, &Transaction)], height: u32) {
+		let mut reentered = true;
+		while reentered {
+			let matched_indexes: HashSet<_> = self.chain_monitor.filter_block(header, txdata).into_iter().collect();
+			let matched_txn: Vec<_> = txdata.iter().cloned().filter(|&(index, _)| matched_indexes.contains(&index)).collect();
+			let last_seen = self.chain_monitor.reentered();
+			{
+				let mut monitors = self.monitors.lock().unwrap();
+				for monitor in monitors.values_mut() {
+					let txn_outputs = monitor.block_connected(header, &matched_txn, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+
+					for (ref txid, ref outputs) in txn_outputs {
+						for (idx, output) in outputs.iter().enumerate() {
+							self.chain_monitor.install_watch_outpoint((txid.clone(), idx as u32), &output.script_pubkey);
+						}
```
I like the direction of shifting this logic down into the `SimpleManyChannelMonitor` (SMCM). 

Would the next step towards actually supporting partial blocks be, maybe, returning a list of txs that the SMCM needs more information about? That way, in theory the BlockNotifier, could try to retrieve that additional info from the LN node's block source. If so, or if there's other ideas for better support for partial blocks, could be worth filing an issue.

jkczyz replied:
That's a good question. Your suggestion seems like a reasonable approach. I'd like to see if this could be unified somehow with `ChainWatchInterface` by possibly removing it entirely or moving it (or something like it) out to the caller of `BlockNotifier`.

As it stands, the `reentered` logic isn't very intuitive. Ideally it can be handled at the `BlockSource` level such that the caller of `BlockNotifier` can determine if the `BlockSource` needs to be re-queried if filtered blocks were provided. Then perhaps `BlockNotifier` would only ever send filtered blocks to listeners.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -36,24 +36,38 @@ use std::mem;
-	let mut header = BlockHeader { version: 0x20000000, prev_blockhash: Default::default(), merkle_root: Default::default(), time: 42, bits: 42, nonce: 42 };
-	notifier.block_connected_checked(&header, 1, &[tx; 1], &[chan_id as usize; 1]);
+pub fn confirm_transaction<'a, 'b: 'a>(notifier: &'a chaininterface::BlockNotifierRef<'b>, tx: &Transaction) {
+	let dummy_tx = Transaction { version: 0, lock_time: 0, input: Vec::new(), output: Vec::new() };
```
Previously, this test utility was calling `block_connected_checked` directly with data already filtered by `ChainWatchInterfaceUtil`. I removed that function now that I've pushed `ChainWatchInterface` down into the listener (i.e., `SimpleManyChannelMonitor`), so I can no longer pass filtered data to `BlockNotifier` directly. Instead, I'm essentially padding the block with dummy transactions up to the desired index then adding the passed `tx` to the end of the vector, which has the same effect as before.

FWIW, this won't be needed if `BlockNotifier` supports filtered blocks directly. But as noted in https://github.com/rust-bitcoin/rust-lightning/pull/649#issuecomment-650335341, I've left that for a future improvement. Given our [discussion](https://github.com/rust-bitcoin/rust-lightning/pull/649#discussion_r447133806), we have to iron out some details still.

jkczyz commented:
Updated the PR based on some of our offline discussions. In summary:
- Replaced `ChainWatchInterface` in `NetGraphMsgHandler` with `chain::Access`
- Replaced `ManyChannelMonitor` in `ChannelManager` with `chain::Watch`, which is essentially `ManyChannelMonitor` with the trait and functions renamed
- Replaced `ChainWatchInterface` in `SimpleManyChannelMonitor` by implementing `chain::WatchEventProvider` for signaling back which transactions and outpoints to watch via `chain::WatchEvent`s
- Removed `ManyChannelMonitor`
- Renamed `SimpleManyChannelMonitor` to `ChainMonitor`
- Removed `ChainWatchInterface` and `ChainWatchInterfaceUtil`

I like the naming but am happy to bikeshed on it later. Will clean-up documentation and commit messages once we're happy with it.

I did not remove `BlockNotifier` and `ChainListener` (yet?). However, I did add a `connect_block` test utility for handling functionality that could be covered by `BlockNotifier` assuming it can interact with a block source for processing the new events. There appears to be a dependency between the order in which the listeners are notified of blocks in some tests. I'd like to understand this further and determine if it can be removed.

Also, since `ChannelMonitor`s can return outpoints that were already signaled, `ChainWatchedUtil` is still needed but is used directly by `ChainMonitor` now for managing event production. This behavior is encapsulated by `WatchEventQueue` in `channelmonitor.rs`.

Block filtering is still inside what is now `ChainMonitor`. I'd like to explore how this might be changed, which may affect what the `BlockListener` interface (or what becomes of it) will look like.

I suspect we'll parameterize `ChainMonitor` by the forthcoming data persistence trait. Test structs may be able to use a custom parameterization instead of wrapping `ChainMonitor`.

jkczyz commented:
Updated PR based on most recent offline discussions. In summary:

- Parameterized `ChainMonitor` by a new `chain::Notify` trait instead of using events
- Removed `BlockNotifier` and `ChainListener`
- Added documentation for new traits
- Updated current documentation
- Updated architecture diagram

TheBlueMatt commented on lightning/src/ln/channelmonitor.rs:
```diff
@@ -159,77 +165,174 @@ impl_writeable!(HTLCUpdate, 0, { payment_hash, payment_preimage, source });
+	events: Vec<WatchEvent>,
+}
+
+/// An event indicating on-chain activity to watch for pertaining to a channel.
```
Given we're already calling flush_events reentrant-ly here, does it really make sense to build a bit cache thing? Or is this just to do the filtering inside ChainMonitor (and it will be removed in a followup?)? I'm a bit confused why we can't just drop this now if there's filtering at the callsite in tests.

TheBlueMatt replied:
> preventing duplicate calls to chain::Notify for the same watch events
> determining if any new watch events were generated (and thus what block_connected should return)

Hmm, can't the ChannelMonitor remove the duplicate calls itself since it now has the full monitoring data? Then if there were any returned values you know you need to rescan.

> I see your point regarding filtering. But if I remove filtering from block_connected, I see a few test failures:

Right, 99% those are just the tests being overly strict, which should be fixed. Totally fine to do it separately, I'm just dubious we can't avoid this new struct even without addressing that.

jkczyz replied:
> Hmm, can't the ChannelMonitor remove the duplicate calls itself since it now has the full monitoring data? Then if there were any returned values you know you need to rescan.

Could you clarify what you mean by "since it now has the full monitoring data"? I didn't change `ChannelMonitor` materially.

TheBlueMatt replied:
Right, it's new in that its from Jan of this year in 73dce207dd0ea6c3ac57af3ebb8b87ee03e82c9e.

jkczyz replied:
I was able to remove `WatchEventCache` -- and also filtering -- and use the full monitoring data without difficulty, but I'm having trouble with the tests. For `test_bump_penalty_txn_on_revoked_htlcs`, the number of broadcasted transactions is reduced by one with the inputs of `node_txn[2]` and `node_txn[3]` combined into a single transaction (`node_txn[2]`). I can adjust [those assertions](https://github.com/rust-bitcoin/rust-lightning/blob/478c481c7f9c8fc10669fb6b9d2f2c16751a432c/lightning/src/ln/functional_tests.rs#L7877-L7880), but the `check_spends` that follows fails. And commenting it out gives an overflow when subtracting in the `fee_1` calculation if replacing `node_txn[3]` with `node_txn[2]` since the output value is larger now with 3 inputs instead of 2.

Any insight into why the two transactions are combined into one and how to go about modifying this test?

TheBlueMatt replied:
Hmm, not immediately, it may be that the filtering in the monitors is broken. Can you push the WIP commit(s) somewhere I can play with?

jkczyz replied:
Change is in 71c76068b5e1f368d52e335e48d704a34878b2c1 and the attempt fix in 5acef01c437b88f7f88ede0ae2323892ac863658.

TheBlueMatt replied:
Looks like its claiming a third (revoked) output - https://github.com/TheBlueMatt/rust-lightning/commits/2020-08-check-spends-fees calculates the fees more robustly and uses them for this specific bug. I'm not actually sure *why* its now spending the commitment-tx revoked output in addition to the HTLC revoked outputs, though. Maybe @ariard can shed some light?

TheBlueMatt commented on lightning/src/ln/channelmonitor.rs:
```diff
@@ -149,8 +155,8 @@ pub struct HTLCUpdate {
 
-/// A simple implementation of a ManyChannelMonitor and ChainListener. Can be used to create a
-/// watchtower or watch our own channels.
+/// A simple implementation of a [`chain::Watch`]. Can be used to create a watchtower or to watch
```
This seems like a strange place for this now - if Watch is in chain, why is the sample implementation in ln? Also likely worth noting that the watchtower stuff no longer makes sense in the comments here, which probably *also* implies we should drop the whole weird Key templating - the original thinking was that you'd Key on the OutPoint for local use and a (PublicKey, u32) for remote use where the originator signs updates with a PublicKey. Since ChannelMonitor is now never a watchtower, the Key differentiation makes no sense and we should probably go to just being a chain::Watch sample implementation.

TheBlueMatt replied:
> Good question. I see it less as a sample implementation and more as a typically (re-)used component that also implements chain::Watch.

That doesn't really change the meat of the issue, though, no? If its really just a `chain::Watch` that you're supposed to use, why isn't it in `chain`, and why is it in `ln`?

> can I assume this outside the scope of this PR?

Sure, you can do it later, I don't feel strongly either way.

jkczyz replied:
> Also likely worth noting that the watchtower stuff no longer makes sense in the comments here, which probably _also_ implies we should drop the whole weird Key templating - the original thinking was that you'd Key on the OutPoint for local use and a (PublicKey, u32) for remote use where the originator signs updates with a PublicKey. Since ChannelMonitor is now never a watchtower, the Key differentiation makes no sense and we should probably go to just being a chain::Watch sample implementation.

Actually, I'll need some clarification to better understand the watchtower case. What was the `u32` meant to signify in the key?

Also, what is meant by "ChannelMonitor is now never a watchtower"? Did you mean `ChainMonitor`?

TheBlueMatt replied:
No, up until (I believe) this year, ChannelMonitor contained an enum that allowed it to "be in watchtower mode" (ie not have private keys), but it was never fully implemented, so removed. In that context, it would make sense to store ChannelMonitors indexed by "public key of user who signed the update" and some channel id per user.

TheBlueMatt commented on lightning/src/chain/mod.rs:
```diff
@@ -1,5 +1,100 @@
+	/// [`get_outputs_to_watch`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.get_outputs_to_watch
+	/// [`block_connected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_connected
+	/// [`block_disconnected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
```
Hmm, this just feels like such strange placement - we're returning a unitary enum type from ln in chain. Maybe move ChannelMonitorUpdateErr here or just move all of the on-chain stuff from ln::channelmonitor to chain::channelmonitor?

jkczyz replied:
Yeah, I was thinking the same. Leaning towards moving to `chain::channelmonitor`. How are we distinguishing between what goes in `chain` vs `ln`?

Any other PRs to land before making the move? May be worth waiting for a follow-up if so.

TheBlueMatt replied:
I dont think there's any ongoing work on this, but historically `ln` was "everything", and there wasn't anything else. Moving all of the channelmonitor stuff to `chain` somehow sounds good to me, and it would clean up this PR a bit, I think, especially to effectively move all of its changes into chain.

jkczyz replied:
SGTM. Just to be clear, this is largely just moving the entire `channelmonitor.rs` file from `ln` to `chain`.

jkczyz replied:
So, it appears this will be more involved since there are a lot of types, functions, and fields with `pub(super)` visibility either in or used by `channelmonitor.rs`. I could make these `pub` or start moving other files. Moving more files isn't very appealing as I think this will have a cascading effect.

TheBlueMatt replied:
Hmm, I mean thats probably an indication of something else - if we want users to be able to implement `chain::Watch` themselves, then anything a default implementation calls probably also needs to be pub.

jkczyz replied:
I think the problem is largely limited to `ChannelMonitor` and related types rather than `ChainMonitor`. Specifically, moving it causes visibility issues in `Channel` and `ChannelManager`. The opposite direction is primarily that `ChannelMonitor` uses `OnchainTxHandler`, though I haven't looked into it exhaustively yet.

TheBlueMatt replied:
Oh, right, hmm. I mean I don't think having to `s/pub(super)/pub(struct)/` is a reason to not move it, though, no?

jkczyz replied:
I'm not arguing to not move it but rather trying to figure out the best approach to resolve visibility without moving more files.
- `pub(<struct>)` is not valid visibility
- `pub(in crate::chain::channelmonitor)` gives "visibilities can only be restricted to ancestor modules"
- `pub(crate)` works
- `pub` works

Leaning towards `pub(crate)` over `pub` so as to not open everything to anyone using the crate. Let me know what you prefer or if I'm missing something.

TheBlueMatt replied:
Err, right, i meant pub(crate). I think that's generally fine.

TheBlueMatt commented on lightning/src/chain/mod.rs:
```diff
@@ -1,5 +1,101 @@
+	/// [`block_disconnected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Updates a channel identified by `funding_txo` by applying `update` to its monitor.
```
Right, well the big comment over ChannelMonitorUpdateErr still exists, so its documented. We should probably move that and merge it into the documentation on `Watch`, IMO.

ariard commented:
@jkczyz 

See this [branch](https://github.com/ariard/rust-lightning/tree/2020-08-fix-remove-chainwatch) for fixing tests and commit message for laying out the reason. To resume, this refactoring doesn't introduce a bug, but ccbc596 by switching utilities test didn't preserve the step by step monitoring scanning of package of transactions, thus triggering OnchainTxHandler aggregation logic more aggressively and diminishing the number of broadcast.

jkczyz commented:
> @jkczyz
> 
> See this [branch](https://github.com/ariard/rust-lightning/tree/2020-08-fix-remove-chainwatch) for fixing tests and commit message for laying out the reason. To resume, this refactoring doesn't introduce a bug, but [ccbc596](https://github.com/rust-bitcoin/rust-lightning/commit/ccbc5968e4948555c5fe23bfe4e2386d9e87e928) by switching utilities test didn't preserve the step by step monitoring scanning of package of transactions, thus triggering OnchainTxHandler aggregation logic more aggressively and diminishing the number of broadcast.

Hmm... so I managed to fix this yesterday but without modifying the tests. I found two problems:

1. I wasn't filtering transactions in `ChannelMonitor`
2. The `outputs_to_watch` did not include the funding txo.

Also, when debugging I noticed that the newly broadcasted transactions did not look valid. They seemed to double-spend some inputs. But with the above fixes this was no longer a problem.

TheBlueMatt commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,101 @@
+/// maintaining a set of monitors such that they can be updated accordingly as channel state changes
+/// and HTLCs are resolved on chain. See method documentation for specific requirements.
+///
+/// TODO: Add documentation about persisting monitors.
```
Can we fill this out now?

jkczyz replied:
Done.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1870,17 +1835,43 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
+
+	/// Determines if any HTLCs have been resolved on chain in the connected block.
+	///
+	/// Returns any new outputs to watch from `txdata`. After called, these are also included in
```
Hmm, I know its not your bug, but these are looking increasingly confusing without a struct that clearly defines what they mean (which I suppose was somewhat the case when it was a function to register watch outputs with documentation).

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -160,119 +164,138 @@ pub enum MonitorEvent {
 	/// interface with which to register to receive notifications.
-	pub fn new(chain_monitor: C, broadcaster: T, logger: L, feeest: F) -> SimpleManyChannelMonitor<Key, ChanSigner, T, F, L, C> {
-		let res = SimpleManyChannelMonitor {
+	pub fn new(chain_source: Option<C>, broadcaster: T, logger: L, feeest: F) -> Self {
```
This needs docs for when/why you'd provided a chain_source and when not.

TheBlueMatt commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,101 @@
+	fn release_pending_monitor_events(&self) -> Vec<MonitorEvent>;
+}
+
+/// The `Notify` trait defines behavior for indicating chain activity of interest pertaining to
```
I'm not sure about the "Notify" name, but, more broadly, should we endeavor to have ChainMonitor::block_connected just return the to-be-registered bits instead? I'm still not convinced its not a better design by avoiding reentrancy, but maybe I forgot some part of the conversation?

TheBlueMatt reviewed (commented):
Sorry this took so long to get to. Mostly a few high-level questions still but I really love where this ended up for the most part.

valentinewallace commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -160,119 +164,138 @@ pub enum MonitorEvent {
 }
 
-/// Simple structure send back by ManyChannelMonitor in case of HTLC detected onchain from a
+/// Simple structure send back by `chain::Watch` in case of HTLC detected onchain from a
```
nit: update comment: "Simple structure *sent* back by `chain::Watch` when an HTLC is detected onchain from a forwarding channel. This info is needed to update the corresponding HTLC in the originating channel."

valentinewallace commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -282,29 +305,29 @@ impl<Key : Send + cmp::Eq + hash::Hash + 'static, ChanSigner: ChannelKeys, T: De
 	}
 
-	fn get_and_clear_pending_monitor_events(&self) -> Vec<MonitorEvent> {
+	fn release_pending_monitor_events(&self) -> Vec<MonitorEvent> {
```
Love this rename

valentinewallace commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1991,10 +1988,10 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 	}
 
-	fn would_broadcast_at_height<L: Deref>(&self, height: u32, logger: &L) -> bool where L::Target: Logger {
+	pub(crate) fn would_broadcast_at_height<L: Deref>(&self, height: u32, logger: &L) -> bool where L::Target: Logger {
```
Tests pass for me when this is private

valentinewallace reviewed (commented):
Partial review, this is looking good to me though!

jkczyz commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,41 @@
+}
+
+/// An event indicating on-chain activity to watch for pertaining to a channel.
+pub enum WatchEvent {
```
Likely will be very difficult to do by squashing as it will make the resulting commit not very atomic. Will see if I can move some of the commits around but I have a feeling it will be ugly.

Would it be any better if I moved the trait and enum to `channelmonitor.rs`?

valentinewallace commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -160,119 +164,143 @@ pub enum MonitorEvent {
+	///
+	/// When an optional chain source implementing [`chain::Filter`] is provided, the chain monitor
+	/// will call back to it indicating transactions and outputs of interest. This allows users to
+	/// pre-filter blocks or only fetch blocks matching a compact filter.
```
Good docs :+1:  Maybe add a sentence: "If no chain source is provided, full blocks must be provided on `block_connected` calls."

valentinewallace commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,103 @@
+/// The `Watch` trait defines behavior for watching on-chain activity pertaining to channels as
+/// blocks are connected and disconnected.
+///
+/// Each channel is associated with a [`ChannelMonitor`]. Implementations are responsible for
```
optional nit (maybe too obvious): Implementations of this trait are responsible...

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -3313,7 +3313,7 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 	/// May return some HTLCs (and their payment_hash) which have timed out and should be failed
 	/// back.
-	pub fn block_connected(&mut self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[usize]) -> Result<(Option<msgs::FundingLocked>, Vec<(HTLCSource, PaymentHash)>), msgs::ErrorMessage> {
+	pub fn block_connected(&mut self, header: &BlockHeader, txdata: &[(usize, &Transaction)], height: u32) -> Result<(Option<msgs::FundingLocked>, Vec<(HTLCSource, PaymentHash)>), msgs::ErrorMessage> {
```
Could this use the new `TransactionData` thing? :)

valentinewallace commented on lightning/src/routing/network_graph.rs:
```diff
@@ -127,29 +129,34 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+					Err(chain::AccessError::UnknownTx) => {
+						return Err(LightningError{err: "Channel announced without corresponding UTXO entry".to_owned(), action: ErrorAction::IgnoreError});
+					},
+					// format!("Channel announced on an unknown chain ({})", msg.contents.chain_hash.encode().to_hex())
```
nit

jkczyz reviewed (commented):
Addressed comments.

jkczyz commented:
Rebased on master. I didn't update `lightning-c-bindings` as `genbindings.sh` failed likely because of #678. May try to see how easy it would be to fix that issue. Otherwise could do some import renames for `chain::Watch` et al.

jkczyz commented:
Looks like the type alias is also problematic:

```
Imported ChainMonitor as ChainMonitor
    Type chain::transaction::TransactionData (ref: true) unresolvable in C
thread 'main' panicked at 'assertion failed: self.write_c_type_intern(generics, w, t, false, false, ptr_for_ref)', src/types.rs:2020:9
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: std::panicking::rust_panic_with_hook
   6: std::panicking::begin_panic
   7: c_bindings_gen::types::TypeResolver::write_c_type
   8: c_bindings_gen::blocks::write_method_params
   9: c_bindings_gen::writeln_impl
  10: c_bindings_gen::convert_file
  11: c_bindings_gen::convert_file
  12: c_bindings_gen::convert_file
  13: c_bindings_gen::main
  14: std::rt::lang_start::{{closure}}
  15: std::rt::lang_start_internal
  16: std::rt::lang_start
  17: main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3173,8 +3170,11 @@ impl<ChanSigner: ChannelKeys, M: Deref + Sync + Send, T: Deref + Sync + Send, K:
+	///
+	/// If necessary, the channel may be force-closed without letting the counterparty participate
+	/// in the shutdown.
+	pub fn block_disconnected(&self, header: &BlockHeader, _: u32) {
```
Probably worth dropping the dummy parameter now that we don't have to conform to a trait, no?

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1875,17 +1839,43 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
+	/// Determines if any HTLCs have been resolved on chain in the connected block.
+	///
+	/// Returns any new outputs to watch from `txdata`. After called, these are also included in
+	/// [`get_outputs_to_watch`].
```
This definitely needs a discussion of the requirements for rescan (that you have to do so before connecting the next block, that it is ok to pass channel-provided updates to the monitor in between), and how you can avoid them (that you need not care whatsoever if you provide the full block data, which should be default if you have it, or if you provide the full set of txn that were in the block which spent any previously-matched txn).

TheBlueMatt replied:
> Can the rescan be avoided given the block is immediately filtered at the start of this function?

Oof, right, that sounds like a bug on its own (we should do the trick of "if it spends something we already matched, it also matches", I think, not sure if there's anything more efficient) :).

> Here, by rescan, I assume you mean call block_connected again with the same block (possibly re-fetched if you don't have full txdata).

Yes, that you (should be able to) avoid calling it again with the block's data refreshed if you provide full block data (or match anything which spends a previous match).

> Not sure if I'm parsing this correctly. Could you confirm my understanding?

ie "User filters block with the match criteria broadened to 'anything that matches or anything which spends a transaction in the same block which matched'".

jkczyz replied:
> Oof, right, that sounds like a bug on its own (we should do the trick of "if it spends something we already matched, it also matches", I think, not sure if there's anything more efficient) :).

@TheBlueMatt I attempted implementing this but am getting some test failures. Here's the branch:

https://github.com/jkczyz/rust-lightning/tree/2020-06-refactor-chain-listener-lazy-filter

Specifically, see: https://github.com/jkczyz/rust-lightning/commit/c4aad5b7789c533c31b95d819bd72cb354190360.

Does this look like the right approach? Could the failures be related to the problem that @ariard noted earlier: https://github.com/rust-bitcoin/rust-lightning/pull/649#issuecomment-683500920? Or is it worth investigating whether the tests might be fragile?

TheBlueMatt replied:
Hmm, I was thinking more like this, though it also fails tests, but seemingly fewer tests fail. May be worth playing with that, but I think in either case its likely the tests just being too specific.

```
@@ -1878,7 +1878,20 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
                      F::Target: FeeEstimator,
                                        L::Target: Logger,
        {
-               let txn_matched = &txdata.iter().filter(|&&(_, tx)| self.does_tx_match(tx)).map(|e| *e).collect::<Vec<_>>();
+               let mut match_set = HashSet::new();
+               let txn_matched = &txdata.iter().filter(|&&(_, tx)| {
+                       let mut matches = self.does_tx_match(tx);
+                       for inp in tx.input.iter() {
+                               if matches { break; }
+                               if match_set.contains(&inp.previous_output.txid) {
+                                       matches = true;
+                               }
+                       }
+                       if matches {
+                               match_set.insert(tx.txid());
+                       }
+                       matches
+               }).map(|e| *e).collect::<Vec<_>>();
                for &(_, tx) in txn_matched {
                        let mut output_val = 0;
                        for out in tx.output.iter() {
```

jkczyz replied:
@TheBlueMatt The failing tests with your solution are the same tests that failed before I added the filter step. I tried adding the test fixes given by @ariard in https://github.com/ariard/rust-lightning/commit/b893cbfdd88391208dca62ab1fe872a7406ff661, and they do indeed result in the tests passing.

However, I'm not sure I understand why this is considered a viable fix. It breaks the transactions across two blocks, but what would prevent a peer from getting the transactions into a single block?

TheBlueMatt replied:
Huh, right, yea, I definitely agree that that's a bogus fix - that seems more like adapting the test to the behavior rather than testing.

TheBlueMatt commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,105 @@
+///
+/// This is useful in order to have a [`Watch`] implementation convey to a chain source which
+/// transactions to be notified of. Notifying may take the form of pre-filtering blocks or, in the
+/// case of [BIP 157]/[BIP 158], only fetching a block if the compact filter matches.
```
+Note that if you are fetching the full blockchain, this trait is likely implemented with empty methods.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -159,121 +163,139 @@ pub enum MonitorEvent {
+	/// be returned by [`chain::Watch::release_pending_htlc_updates`].
+	///
+	/// Calls back to [`chain::Filter`] if any monitor indicated new outputs to watch, returning
+	/// `true` if so.
```
+ Note that in such a case you likely need to refetch the filtered block with updated transactions, unless you provided the full blocks' transaction data.

TheBlueMatt commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,105 @@
+///
+/// [`Watch`]: trait.Watch.html
+/// [BIP 157]: https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki
+/// [BIP 158]: https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki
```
This needs some discussion of timing requirements (ie that once the function returns, any blocks connected thereafter must contain the transaction), and a discussion about reentrancy here (also that these functions should *not* block and should queue the new monitoring data if it needs to be sent to a remote server, returning TemporaryFailure from the chain::Watch functions if you want to block, or simply rescanning if you do not).

TheBlueMatt replied:
>  I'm not sure I follow the parts about returning TemporaryFailure from chain::Watch

I'm thinking this - lets say someone wants everything to be async (great), and they have a remote server they need to send the new data to watch. They want to return right away (cause everything is async, cool), but haven't received confirmation from the server that the data has been registered. If we're getting new data to watch from a call to block_connected, presumably they can figure out some other way to be compliant. If we're getting new data to watch from a new monitor, what they could do, instead of complying with the requirement as stated, is return `TemporaryFailure` from `watch_channel` and, once they get confirmation from the server, call `ChannelManager::channel_monitor_updated`.

jkczyz replied:
Ah, makes sense. Thanks for explaining. I've updated the documentation but left out anything about rescans until that topic has been resolved.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3054,17 +3051,17 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
+        L::Target: Logger,
 {
-	fn block_connected(&self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[usize]) {
+	/// Updates channel state based on transactions seen in a connected block.
```
Note that unlike chain::ChannelMonitor, this function does *not* need any kind of block rescan logic.

TheBlueMatt reviewed (commented):
Few comments, I'm working on updates to the bindings generator for this pr now.

julianknutsen commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -44,51 +44,76 @@ use std::mem;
+
+pub fn connect_block<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, block: &Block, height: u32) {
+	let txdata: Vec<_> = block.txdata.iter().enumerate().collect();
+	while node.chain_monitor.chain_monitor.block_connected(&block.header, &txdata, height) {}
```
Can you explain when this would return true twice in a row? Is it a concurrency situation or something else?

julianknutsen commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -60,7 +64,7 @@ use std::io::Error;
 #[must_use]
 pub struct ChannelMonitorUpdate {
-	pub(super) updates: Vec<ChannelMonitorUpdateStep>,
+	pub(crate) updates: Vec<ChannelMonitorUpdateStep>,
```
Is this the final visibility? Or is the final module structure going to encapsulate this again?

jkczyz replied:
I'm not certain what the final visibility will look like. This is used by `channel.rs`, so it needs to be `pub(crate)` for now and likely foreseeably so. We may be able to provide an interface that encapsulates it again, however.

julianknutsen commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1210,6 +1210,10 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 			htlc_outputs: Vec::new(), // There are never any HTLCs in the initial commitment transactions
 		};
+
+		let mut outputs_to_watch = HashMap::new();
```
Seems like a regression test should go with the fix. How does the test infrastructure look around this code?

jkczyz replied:
IIUC, `outputs_to_watch` was added but not used in such a way that required the funding txo. When I began using it in such a manner, a few tests broke. This commit is what I used to fix the tests, though I placed the commit prior to the breaking change.

julianknutsen replied:
I see. I just didn't see the cause and effect in my pass through. Thanks for the background.

julianknutsen commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1973,21 +1965,25 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
-			self.outputs_to_watch.insert(txid.clone(), output_scripts.iter().map(|o| o.script_pubkey.clone()).collect());
-		}
-
+		watch_outputs.retain(|&(ref txid, ref txouts)| {
```
A comment similar to the commit message explaining this block help readability.

jkczyz replied:
Good point! Done.

julianknutsen commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1875,17 +1839,43 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
-	/// on-chain.
-	fn block_connected<B: Deref, F: Deref, L: Deref>(&mut self, txn_matched: &[&Transaction], height: u32, block_hash: &BlockHash, broadcaster: B, fee_estimator: F, logger: L)-> Vec<(Txid, Vec<TxOut>)>
+	/// Checks if a given transaction matches the funding transaction or spends any watched outputs.
+	fn does_tx_match(&self, tx: &Transaction) -> bool {
```
fuction name is ambiguous given the comment. `does_tx_match_funding_or_watched()`?

jkczyz replied:
I'm up for renaming, but what part is ambiguous? Would `is_relevant_tx` be sufficient or too vague? I was hoping to opt for a shorter names given this is passed to `filter`.

julianknutsen replied:
`is_relevant_tx` seems great like an improvement, but feel free to throw this in the bikeshed pile and move on.

julianknutsen reviewed (commented):
Moved my comments from 681. Will review changes since Val's last rebase tomorrow.

jkczyz commented on lightning/src/util/test_utils.rs:
```diff
@@ -393,27 +393,32 @@ impl TestKeysInterface {
-		let script = Builder::new().push_opcode(opcodes::OP_TRUE).into_script();
-		Self { utxo_ret: Mutex::new(Ok((script, u64::max_value()))) }
+impl TestChainSource {
+	pub fn new(network: Network) -> Self {
```
Currently, the tests use fake data. Parameterizing `Network` makes it simpler to test `get_utxo` but is probably not absolutely necessary.

jkczyz commented on lightning/src/util/test_utils.rs:
```diff
@@ -393,27 +393,32 @@ impl TestKeysInterface {
 
-pub struct TestChainWatcher {
-	pub utxo_ret: Mutex<Result<(Script, u64), ChainError>>,
+pub struct TestChainSource {
```
Noted!

jkczyz commented on lightning/src/chain/chaininterface.rs:
```diff
@@ -378,77 +378,62 @@ impl ChainWatchInterfaceUtil {
 
+	struct TestChainListener(u8);
+
+	impl ChainListener for TestChainListener {
```
That's a good suggestion. Would rather punt on this for now given the size of the PR and the fact that these tests were deleted later.

Having `Test` as part of the name seems redundant/verbose and differs from the linked article. Is this primarily to aid auto-complete?

julianknutsen replied:
The article was mostly just to point to a reference of how the names (Spy/Mock/Stub) map to test contracts and less about exact naming. I like putting `TestStub` more than `Stub` since there are a lot of devs that may not have been introduced to the concept and test vs. real objects and it help start the conversation and is easily googleable for resources.

I like suffixes more than prefixes for the autocomplete and in my opinion it just reads a little easier when you are looking through a test trying to understand which interfaces are involved.

Definitely a punt, just wanted to sync up about it.

jkczyz reviewed (commented):
Addressed all comments, though there are a few open questions for @TheBlueMatt in a few of them.

julianknutsen commented on lightning/src/routing/network_graph.rs:
```diff
@@ -127,29 +129,33 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
-			},
-			Err(ChainError::NotSupported) => {
+		let utxo_value = match &self.chain_access {
+			&None => {
```
Totally agree. The change to 1.30 makes this possible and just wanted to spread the word.

julianknutsen commented:
Nothing of mine was blocking to begin with, but all of my questions have been addressed. Thanks, Jeff.

TheBlueMatt commented:
For my own sanity bindings-wise (and because I think it'd be better) can you move ChannelMonitor to a separate file from ChainMonitor? If you don't want to grow the size of this PR I'll do it separately in a bindings update.

jkczyz commented:
> For my own sanity bindings-wise (and because I think it'd be better) can you move ChannelMonitor to a separate file from ChainMonitor? If you don't want to grow the size of this PR I'll do it separately in a bindings update.

Sure thing, I don't mind.

jkczyz commented:
I've split out `ChainMonitor` into a separate module. Also pushed commits related to discussion in https://github.com/rust-bitcoin/rust-lightning/pull/649#discussion_r489813539.

ariard commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,115 @@
+///
+/// This is useful in order to have a [`Watch`] implementation convey to a chain source which
+/// transactions to be notified of. Notifying may take the form of pre-filtering blocks or, in the
+/// case of [BIP 157]/[BIP 158], only fetching a block if the compact filter matches. After a
```
`in the case of a lightclient protocol such as...` Electrum is still widely used as of today.

ariard commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,115 @@
+pub trait Filter: Send + Sync {
+	/// Registers interest in a transaction with `txid` and having an output with `script_pubkey` as
+	/// a spending condition.
+	fn register_tx(&self, txid: &Txid, script_pubkey: &Script);
```
As a follow-up, I think we can simplify this trait by registering funding_tx by one of its output. Further the script_pubkey is worthless here as mutation of it would break the txid.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3173,8 +3174,11 @@ impl<ChanSigner: ChannelKeys, M: Deref + Sync + Send, T: Deref + Sync + Send, K:
+	/// Updates channel state based on a disconnected block.
+	///
+	/// If necessary, the channel may be force-closed without letting the counterparty participate
+	/// in the shutdown.
```
What does this mean ? IIRC the only case where you may have to force-close a channel due to disconnection is because the funding is disconnect, so it can't imply interacting with the counterparty as there is not mutual closing tx to agree on ?

ariard replied:
Right we mean the same thing "without interacting with the counterparty" is clearer as `shutdown` is a well-defined LN message :) But I'm not even sure we can qualify this of a force-close like when we broadcast the commitment. We just forget about the channel.

ariard replied:
We return a `ChannelMonitorUpdateStep::ChannelForceClosed` which is going to be send to chain_monitor. The first and (last) commitment transaction will try to be broadcast, which should result in a failure as the funding transaction isn't confirmed. I don't think we should worry about this outcome.

jkczyz replied:
Do you have a specific recommendation here? (e.g., remove the sentence entirely)

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -127,29 +129,33 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+			&Some(ref chain_access) => {
+				match chain_access.get_utxo(&msg.contents.chain_hash, msg.contents.short_channel_id) {
+					Ok(TxOut { value, script_pubkey }) => {
+						let expected_script = Builder::new().push_opcode(opcodes::all::OP_PUSHNUM_2)
```
An easy follow-up `get_funding_transaction_redeemscript()` to improve readability.

ariard reviewed (commented):
Apart of the filtering for the funding transaction, see comment and updating few comments the rest can be bookmarked as follow-ups, I think.

ariard commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -263,29 +282,29 @@ impl<Key : Send + cmp::Eq + hash::Hash + 'static, ChanSigner: ChannelKeys, T: De
 
-	fn update_monitor(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.update_monitor_by_key(funding_txo, update) {
+	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
```
Sorry that's me. `ChannelMonitor` doesn't "update" channel in the sense of provide signatures for the counterparty for a newer commitment transaction, it "reacts" on onchain events. 

I'm aimed to have a really clear function naming to underscore well the master-slave relation between `chain::Watch` caller and the implementation here.

ariard commented on lightning/src/chain/mod.rs:
```diff
@@ -1,5 +1,101 @@
+	/// is unknown.
+	///
+	/// [`short_channel_id`]: https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#definition-of-short_channel_id
+	fn get_utxo(&self, genesis_hash: &BlockHash, short_channel_id: u64) -> Result<TxOut, AccessError>;
```
Yes sure

ariard commented on lightning/src/chain/mod.rs:
```diff
@@ -1,5 +1,101 @@
+/// blocks are connected and disconnected.
+///
+/// Each channel is associated with a [`ChannelMonitor`]. Implementations are responsible for
+/// maintaining a set of monitors and updating them accordingly as channel state changes and HTLCs
```
Sounds good to me, we can have big documentary-only follow-up to clarify latter.

ariard commented:
@jkczyz 

> The outputs_to_watch did not include the funding txo.

IIRC we don't use the `funding_txo` isn't used by `ChannelMonitor::block_connected`, as of today we don't need to match the funding tx to trigger the ChannelMonitor processing.

>  Also, when debugging I noticed that the newly broadcasted transactions did not look valid. They seemed to double-spend some inputs. But with the above fixes this was no longer a problem.

This is tie to the ongoing discussion on the rescan. Also, and this is will be always true, we might reschedule spend with higher-feerate and thus a different txid. A lawful double-spend according to RBF (BIP125)

jkczyz commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1875,17 +1661,62 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
-	fn block_connected<B: Deref, F: Deref, L: Deref>(&mut self, txn_matched: &[&Transaction], height: u32, block_hash: &BlockHash, broadcaster: B, fee_estimator: F, logger: L)-> Vec<(Txid, Vec<TxOut>)>
+	/// Checks if a given transaction matches the funding transaction or spends any watched outputs.
+	fn is_relevant_tx(&self, tx: &Transaction) -> bool {
+		let funding_txo = self.get_funding_txo();
```
I had been largely just replicating the behavior of `ChainWatchedUtil`. However, you're right that it doesn't actually matter. Removed.

jkczyz commented:
Accidentally sent some comments early. Will push once I address the rest.

jkczyz reviewed (commented):
Addressed most comments or left a response if needed more clarity.

> @jkczyz
> 
> > The outputs_to_watch did not include the funding txo.
> 
> IIRC we don't use the `funding_txo` isn't used by `ChannelMonitor::block_connected`, as of today we don't need to match the funding tx to trigger the ChannelMonitor processing.
> 

It is used when filtering the passed `txdata` for transactions of interest now. Otherwise, many tests fail.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1875,17 +1661,62 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
+
+		for input in tx.input.iter() {
+			for (txid, outputs) in self.get_outputs_to_watch().iter() {
+				for (idx, _script_pubkey) in outputs.iter().enumerate() {
```
Hmm, if we want to do that we should probably do it comprehensively and have a thought-out threat model. Can you open an issue for this and we'll do it in a future PR? Maybe in some kind of wrapper function, though it seems awkward to force users into us revalidating transactions that they may have already validated.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1875,17 +1661,62 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
+	///
+	/// Returns any new outputs to watch from `txdata`; after called, these are also included in
+	/// [`get_outputs_to_watch`]. If new outputs were returned, callers must rescan the block before
+	/// connecting the next block. However, processing channel-provided updates between rescans is
```
Right, well as of #715 we should drop any references to "rescans" in the codebase :tada:

valentinewallace commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,117 @@
+/// completion. If an update fails with a [`PermanentFailure`], then it must immediately shut down
+/// without taking any further action such as persisting the current state.
+///
+/// If an implementation maintains multiple instances of a channel's monitor (e.g., by storing
+/// backup copies), then it must ensure that updates are applied across all instances. Otherwise, it
```
This contradicts a discussion on the 681: https://github.com/rust-bitcoin/rust-lightning/pull/681#discussion_r484259489 where Antoine/devrandom concluded that the majority is safe for backups. Not sure whose should be updated though (I don't feel strongly about majority vs. all instances for v1).

ariard replied:
It's really a two-layer deployment : 

* at the monitor-level, you need consensus of all your monitor instance to move the off-chain state forward
* at the data-level, you need only 1 not -defaulting shard, a majority means a higher level of reliability

You may have Alice, the off-chain channel manager requiring from Bob, Dave and Caroll, its monitors, to acknowledge the state update and then Bob may ask internally to Eve and Fanny its shards to store the data before replying to Alice.

Yes this starts to be a fancy setup but this is what you want to pour confidently double-digit of BTCs in your channels :) ?

valentinewallace commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,117 @@
+/// transaction or an output has been registered, subsequent block retrievals from the chain source
+/// must not exclude any transactions matching the new criteria.
+///
+/// Note that use as part of a [`Watch`] implementation involves reentrancy. Therefore, these
```
nit: maybe specify, reentrancy into user code

TheBlueMatt replied:
Yea, that plus its a bit confusing to have ChannelMonitor::block_connected end up calling things in your Filter, notably it would be easy to have lock inversions registering a tx while some block_connected caller is looking at the tx set (for some reason, though in the latest API they probably shouldnt be, but whatever).

valentinewallace commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,117 @@
+/// Note that use as part of a [`Watch`] implementation involves reentrancy. Therefore, these
+/// functions should not block on I/O. Implementations should instead queue the newly monitored data
+/// to be processed later. This may also require returning [`TemporaryFailure`] from `Watch`
+/// functions in order to block until the data has been processed.
```
Nice, would it be redundant to mention to call `ChannelManager::channel_monitor_updated` when data is done processing? or have a pointer to more specific instructions on this situation?

valentinewallace reviewed (commented):
Reviewed the docs. I kinda braindumped improvement ideas, but I think the docs look pretty good as-is :)

ariard commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,115 @@
+
+/// The `Access` trait defines behavior for accessing chain data and state, such as blocks and
+/// UTXOs.
+pub trait Access: Send + Sync {
```
> Seems this amounts to don't trust data from a source you don't trust.

This should be underscored as a we consume this data downstream for trusted operations (e.g validating our network view). I'm not sure that developers of implementors will understand this as clearly as we do. Though maybe implementors trade-off should be documented in their own high-level documentations ?

ariard commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,115 @@
+///
+/// Implementations **must** ensure that updates are successfully applied and persisted upon method
+/// completion. If an update fails with a [`PermanentFailure`], then it must immediately shut down
+/// without taking any further action such as persisting the current state.
```
Right and I'm not sure if the previous documentation is fault-tolerant proof. Let's keep as it is for now, I'll think on it further.

ariard commented:
> It is used when filtering the passed txdata for transactions of interest now. Otherwise, many tests fail.

You're right in the new model, it's registration isn't done anymore by ChainMonitor for the account of one its ChannelMonitor but directly by the ChannelMonitor itself as a `get_outputs_to_watch` element. In the previous ChannelMonitor::block_connected, we didn't watch it no more return it as a new output to watch. That's the source of my confusion :)

ariard commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,219 @@
+//! to update on-chain [`ChannelMonitor`]s accordingly. If any on-chain events need further
+//! processing, it will make those available as [`MonitorEvent`]s to be consumed.
+//!
+//! `ChainMonitor` may be used directly to monitor channels locally or as a part of a distributed
+//! setup to monitor channels remotely. In the latter case, a custom `chain::Watch` implementation
+//! would be responsible for routing each update to a remote server and for retrieving monitor
+//! events. The remote server would make use of `ChainMonitor` for block processing and for
+//! servicing `ChannelMonitor` updates from the client.
```
I know, this is https://github.com/rust-bitcoin/rust-lightning/issues/604. Writing a blog post is a good call :)

ariard commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,219 @@
+	/// [`ChannelMonitor::block_connected`] for details. Any HTLCs that were resolved on chain will
+	/// be returned by [`chain::Watch::release_pending_monitor_events`].
+	///
+	/// Calls back to [`chain::Filter`] if any monitor indicated new outputs to watch, returning
+	/// `true` if so. In this case, if providing pre-filtered blocks, the caller should re-fetch the
+	/// block to obtain updated `txdata` and recall `block_connected`.
```
Yes we need a clear high-level doc explaining assumptions between Neutrino and Electrum.

ariard commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,219 @@
+
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel based on the disconnected block. See [`ChannelMonitor::block_disconnected`] for
+	/// details.
```
For now our HTLC onchain-to-offchain are delayed by `ANTI_REORG_DELAY`. If we ever reorg so far for now we are just broken (and likely other parts of the Bitcoin ecosystem). Though in the future that would be cool to expose this as configurable option.

ariard commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -159,176 +160,19 @@ pub enum MonitorEvent {
 	CommitmentTxBroadcasted(OutPoint),
```
@TheBlueMatt don't you think we should have also `CounterpartyTxBroadcasted/RevokedTxBroadcasted` to reconciliate offchain state with our on-chain view as soon as we can. For revoked, we should close all our channels with a cheating peer. 

I wonder if we shouldn't just lockdown monitor at any spend of the funding transaction. I'll open an issue.

ariard reviewed (commented):
Still want to review test changes. Feel free to take the better comment suggestions.

jkczyz commented:
Update: I've rebased this post #715 and have squashed the fixup commits. Will address the remaining comments before pushing.

jkczyz commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,219 @@
+{
+	type Keys = ChanSigner;
+
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
```
Removed `pub` from `add_monitor` and `update_monitor` since those are internal. Hopefully that should be sufficient to avoid any user confusion.

jkczyz reviewed (commented):
Addressed the comments or asked for clarification. Let me know if I missed anything.

ariard reviewed (commented):
Code Review ACK  ccac948

I've reviewed that test changes are correct. In fact the hard part was in #715

TheBlueMatt reviewed (commented):
Just read through the docs, but from previous reviews I doubt there's much left there either. This is sooo much nicer than current upstream, so really happy this is getting close.

jkczyz commented on lightning/src/chain/mod.rs:
```diff
@@ -9,6 +9,124 @@
+///
+/// Note that use as part of a [`Watch`] implementation involves reentrancy. Therefore, these
+/// functions should not block on I/O. Implementations should instead queue the newly monitored data
+/// to be processed later. This may also require returning [`TemporaryFailure`] from `Watch`
```
To make sure I understand, when you say:

> should they limit it to when _called from_ a `Watch` function

Is "it" referring to "returning `TemporaryFailure`"?

And if so, are you saying they should only return `TemporaryFailure` from a `Watch` function if:
1. the `Watch` function invocation calls a `Filter` function (e.g. `register_output`) and
2. the `Filter` is waiting on the remote server's response from previously queued outputs to monitor?

If so -- aside from clearing up the docs -- would it be best to have the `Filter` functions signal this back as the return value? That would allow us to implement the logic for returning `TemporaryFailure` in `ChainMonitor`.  Otherwise, we would put the onus on the user to implement their own `Watch` in lieu of using the provided `ChainMonitor`. And such an implementation I'd expect would introduce a strong coupling between the user's `Watch` and `Filter` implementations.

TheBlueMatt replied:
Yes, your understanding is correct - we could let `Filter` return the error, though its somewhat awkward to correctly communicate that this now requires calling `ChannelManager::channel_monitor_updated` which the `update_id` which was contained in the `ChannelMonitorUpdate` object (but presumably wasn't passed into the `Filter`? I suppose we could pass it all the way through but it feels like a layer violation (though of course does seem more usable?).

jkczyz replied:
Ah, I see. Taking a step back, I see the following situation when using `ChainMonitor` as the `chain::Watch`:
- `ChannelManager` calls `ChainMonitor::watch_channel` upon `funding_created`/`funding_signed`
- `ChainMonitor::watch_channel` registers the funding output with the user's `chain::Filter`, which it may queue for processing

Then either:
1. the same situation happens for another channel or
2. the user calls `ChainMonitor::block_connected`, which may register more outputs with the user's `chain::Filter`

There are no other situations where the `chain::Filter` is called from `ChainMonitor`, AFAICT.

Questions:
- Does (1) require any different action other than again returning `TemporaryFailure` unconditionally?
- Would the user wait until the remote server finishes processing before (2), i.e., calling `ChainMonitor::block_connected`?

Assuming the answers are no and yes, respectively, would it be reasonable for `ChainMonitor` to track `ChannelMonitorUpdate::update_id` and provide it through an interface to the user? This would allow the user to pass it to `ChannelManager::channel_monitor_updated` when the remote server has finished processing. Then the user wouldn't need to implement their own `chain::Watch` in this situation.

Anyhow, although this is related, I think I'm getting a little sidetracked. :) Going back to the original discussion on the documentation, I've re-worded the last sentence as:

> Then, in order to block until the data has been processed, any `Watch` invocation that has called the `Filter` must return `TemporaryFailure`.

Does that convey the idea sufficiently?

TheBlueMatt replied:
Right, I think the new wording is fine. To answer your question, I think the answer is either A or B, but both are not required - the user could always Ok() true and just make the server do some kind of rescan dance if there's a race, or the user could always return TemporaryFailure + channel_monitor_updated and then never worry about rescans, but they don't need to do both.

jkczyz commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,226 @@
+	///
+	/// Calls back to [`chain::Filter`] if any monitor indicated new outputs to watch, returning
+	/// `true` if so. In this case, if providing pre-filtered blocks, the caller should re-fetch the
+	/// block to obtain updated `txdata` and re-call `block_connected`.
```
> Luckily this is no longer the way we're going to go (yay tons less complexity)...that said we need to note _somewhere_ that if you do filter a block you must also include any in-block descendants of matched transactions...I presume in the docs for Filter?

Updated `chain::Filter` docs to mention in-block descendants of matched transactions

> Probably means we can simplify the code, too.

By this I take it you mean we no longer need to return a bool and have the caller refetch/rescan? I tried removing the loop from `connect_block` in our functional test utils, but unfortunately doing so causes six tests to fail.

```
failures:
    ln::functional_tests::test_claim_on_remote_revoked_sizeable_push_msat
    ln::functional_tests::test_claim_on_remote_sizeable_push_msat
    ln::functional_tests::test_dynamic_spendable_outputs_local_htlc_timeout_tx
    ln::functional_tests::test_key_derivation_params
    ln::functional_tests::test_static_spendable_outputs_justice_tx_revoked_htlc_success_tx
    ln::functional_tests::test_static_spendable_outputs_timeout_tx
```

Is this an indication that the blocks used in the tests need to include descendant transactions?

TheBlueMatt replied:
Hmm, I can take a look but I'd bet that they're simply assuming reorgs which double-generate transactions needlessly. To get this over the line I'm happy to ignore that for now, but we should fix it before release.

TheBlueMatt replied:
We should still update the docs to note that you do not need to do this (even if the tests do because they're overly restrictive), and we *should* note that you have to include descendants of matched transactions.

jkczyz replied:
Leaving simplification for a follow-up. Re-worded `ChainMonitor::block_connected` docs. The part about not needing to refetch seems a little out of place. Let me know if you have a better suggestion.

jkczyz reviewed (commented):
Made some docs changes from the feedback and left some open questions. Will think on the other unresolved replies.

TheBlueMatt commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,226 @@
+			log_trace!(self.logger, "Got new Channel Monitor for channel {}", log_bytes!(funding_txo.0.to_channel_id()[..]));
+
+			if let Some(ref chain_source) = self.chain_source {
+				chain_source.register_tx(&funding_txo.0.txid, &funding_txo.1);
```
We can drop this now that we always put the funding txo in outputs_to_watch, correct?

jkczyz replied:
Yep, done!

TheBlueMatt replied:
Oops! I was wrong lol. Thanks for following the full changeset through to notice that `register_tx` was now unused, which is wrong :). So the issue here is that `register_output` only implies "spends of the given output", whereas `register_tx` implies "I care about the tx I'm telling you about" - for the funding tx, indeed, we care about spends, but *also* we care about the tx. The tests (obviously, ugh) don't have coverage here, presumably because now we're relying on the in-`ChannelMonitor` filtering to catch bugs (which is almost entirely good news, fewer cases to test), and it does everything except for this one thing, where we care about the ChannelManager getting notified, not the monitor!

jkczyz replied:
Removed the fix-up commit that removed `register_tx`.

ariard replied:
This is still confusing, we have ChainMonitor registering the funding_tx for the account of ChannelManager. A ChannelMonitor is never interested by confirmation of the funding transaction itself, only by its _spends_.

We should either:
* pass a chain::Filter to ChannelManager which would call itself register_tx and ChainMonitor restrains to call register_output on funding
* add a new MonitorEvent::FundingConfirmed returns through `release_pending_monitor_events` and move funding confirmation tracking directly into ChannelMonitor

I think 2nd option is better as you increase fault-tolerance in a distributed deployment.

TheBlueMatt replied:
I dont know if it changes the logic much, but the second options seems like a really great idea if only because we could potentially get rid of block_connected in ChannelManager entirely.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2001,7 +1801,41 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
+	/// Checks if a given transaction spends any watched outputs.
+	fn spends_watched_output(&self, tx: &Transaction) -> bool {
+		for input in tx.input.iter() {
+			for (txid, outputs) in self.get_outputs_to_watch().iter() {
```
This is a HashMap, we don't have to iterate over it :).

jkczyz replied:
Right.. too much literal translation on my part. :P

TheBlueMatt commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,226 @@
+	/// Calls back to [`chain::Filter`] with the funding transaction and outputs to watch.
+	///
+	/// [`chain::Filter`]: ../trait.Filter.html
+	fn add_monitor(&self, outpoint: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), MonitorUpdateError> {
```
This and update_monitor below can just be inlined in the `impl chain::Watch` block, since they're non-pub and only called from there.

TheBlueMatt replied:
Oh, maybe dont do this, since Val did it in 681 and no need to make her rebase.

jkczyz replied:
Yep, was thinking the same. @valentinewallace

TheBlueMatt commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -161,27 +186,23 @@ impl<'a, 'b, 'c> Drop for Node<'a, 'b, 'c> {
 					panic!();
 				}
 			}
-			if chain_watch != *self.chain_monitor {
```
I believe this drop is actually what led to the non-coverage of https://github.com/rust-bitcoin/rust-lightning/pull/649/files#r494694908. If we create a new chain source again and use that when creating the new TestChainMonitor instead of using `self.chain_source` we should be able to get it back easily.

jkczyz replied:
Good point. Stored watched outputs in `TestChainSource` and re-added the check.

TheBlueMatt reviewed (commented):
Pretty much there. Will do another once-over on the docs in a minute or two.

jkczyz commented on lightning/src/chain/chainmonitor.rs:
```diff
@@ -0,0 +1,219 @@
+
+			if let Some(ref chain_source) = self.chain_source {
+				chain_source.register_tx(&funding_txo.0.txid, &funding_txo.1);
+				for (txid, outputs) in monitor.get_outputs_to_watch().iter() {
```
Fixed as per suggested in https://github.com/rust-bitcoin/rust-lightning/pull/649#discussion_r497111743.

TheBlueMatt reviewed (approved):
Needs squash, but I'm happy.

jkczyz commented:
Squashed. Also merged a separate branch for splitting channelmonitor.rs as it made it possible to preserve commit history across both files. Let me know if this is a problem.

ariard reviewed (commented):
Code Review ACK 6cd6816

I've left comments on how we may improve interfaces in the future but for now I think this is okay. I feel we won't agree on the error/async until we start playing with samples.
